### PR TITLE
Add user in script + fixes

### PR DESCRIPTION
**Pull request by parente**
Two additional commits to PR #16 for doc updates, subimage Dockerfile updates, retaining $WORK path compatibility with existing images, and avoiding recursive permission setting on restarts.

@rgbkrk could use another set of eyes on this.

/cc @costerwi

### CODE REVIEW

File: all-spark-notebook/Dockerfile [hunk 1]
```diff
@@ -3,8 +3,6 @@ FROM jupyter/minimal-notebook
 
 MAINTAINER Jupyter Project <jupyter@googlegroups.com>
 
-USER root
```
Comment (rgbkrk) [inline all-spark-notebook/Dockerfile]: Is this assumed to be root at this point? We are running `apt-get` right after, so we might as well be explicit.

**Comment by rgbkrk** [comment]
These are nice global installation setups. :+1:

File: minimal-notebook/notebook.conf [hunk 2]
```diff
@@ -1,18 +1,10 @@
 [program:notebook]
 user=jovyan
```
Comment (rgbkrk) [inline minimal-notebook/notebook.conf]: The minimal notebook specifies `NB_USER` and `environment` below makes use of it. I *would normally* suggest that `NB_USER` then gets propagated, but then a user only has to set `NB_USER` to `root` to elevate permissions.

File: all-spark-notebook/Dockerfile (hunk 1 above)
Comment (parente) [inline all-spark-notebook/Dockerfile]: None of the Dockerfiles do any kind of USER switching anymore after these changes are applied. However, explicit is better than implicit so I'm fine adding a USER root to the beginning of every Dockerfile just to make it clear.

File: minimal-notebook/notebook.conf (hunk 2 above)
Comment (parente) [inline minimal-notebook/notebook.conf]: Would be nice except supervisor doesn't respect env vars in the user field as @costerwi pointed out in PR #20.

**Comment by parente** [comment]
Now has USER root explicitly stated.

**Comment by costerwi** [comment]
These are good changes!  I especially like checking for whether the user was already created.  A few comments:

* We use `$NB_USER` and `jovyan` so interchangeably that it really isn't a variable right now; it would break everything if it were changed.  Should we just replace `$NB_USER` with `jovyan` everywhere to avoid misleading people?
* It's too bad dependent projects need to mount into /home/jovyan/work.  It makes the script awkward and it's more typing for everybody.  Maybe it would make things a little easier if we specify a unique home directory for jovyan with `useradd -d /jovyan ...` ?

**Comment by parente** [comment]
> We use $NB_USER and jovyan so interchangeably that it really isn't a variable right now; it would break everything if it were changed. Should we just replace $NB_USER with jovyan everywhere to avoid misleading people?

Agreed.

> It's too bad dependent projects need to mount into /home/jovyan/work. It makes the script awkward and it's more typing for everybody. Maybe it would make things a little easier if we specify a unique home directory for jovyan with useradd -d /jovyan ... ?

My concern here is people already running `docker run -v /jupyter_notebook_config.py:/home/jovyan/.jupyter/ jupyter/some-notebook-stack:4.0` to volume mount the notebook directory on the host. If we change the path to anything else, the next time they bring pull the 4.0 image, they're going to be in for a surprise. I feel that paths, ports, user names, etc. are part of the "API" of the container.

Really, this is a problem with the versioning scheme on the container images. It would be nice to have tags 4.0_0, 4.0_1, etc. indicating the version of the notebook server and the build of the container that might introduce a breaking change. Unfortunately, the only way to make this happen with automated builds on Docker Hub is to create a new branch for each change like this or new subfolders within the repo. I am interested in solving this problem (issue #15, issue #12), but it might take some more time and thinking to do right.

**Comment by rgbkrk** [comment]
Ok, awesome. Certainly happy with this.

**Comment by rgbkrk** [comment]
> My concern here is people already running docker run -v /jupyter_notebook_config.py:/home/jovyan/.jupyter/ jupyter/some-notebook-stack:4.0 to volume mount their config from the host. If we change the path to anything else, the next time they bring pull the 4.0 image, they're going to be in for a surprise. I feel that paths, ports, user names, etc. are part of the "API" of the container.

That is definitely a difficult balance. Once we unintentionally make a standard, that's... the one people are going to rely on. However, these images haven't been around *that* long. It's the kind of thing where you can post to the mailing list and other comms, hoping to reach some folks before making sweeping changes.

**Comment by parente** [comment]
Would it be OK to continue the path change in a separate issue? It's working in the current notebook-start script, only adds one additional if-block, and is independent of the UID work.

**Comment by rgbkrk** [comment]
Seems fine to me. I'm not of strong opinion here.

**Comment by costerwi** [comment]
This is a good compromise.  Thanks for updating!

**Pull request merged**